### PR TITLE
Create a view for string slicing to avoid extra allocations

### LINKED ISSUE #3736: UnicodeView type
As an optimization, we could create a special `UnicodeViewType` which is functionally identical to a `UnicodeType`, but is a zero-copy view into the storage of another `UnicodeType`.  This would make slicing a zero allocation operation, and also speed up `str.split()` a lot (see #3678 for some discussion there).

Some notes:

* The behavior here would be similar to NumPy, where the view holds a reference count on a buffer object that it may or may not own.  This means a small view of a long string might force the entire string to stick around, increasing memory usage.
* `UnicodeViewType` would have no unboxing method, and boxing would need to copy the data to Python storage (as `UnicodeType` already does), but also decref the underlying string buffer.

### PR DESCRIPTION
**Pull request by ehsantn**
As title. Resolves #3736.

This improves the performance of `split()` significantly in my use case.

### CODE REVIEW

**Comment by seibert** [comment]
Looks good!  Tagging Siu to double check the reference counting logic.  How much is the speedup you see?

**Comment by ehsantn** [comment]
Thanks. It's about 40% for split I think.

File: numba/unicode.py [hunk 1]
```diff
@@ -702,6 +702,41 @@ def _strncpy(dst, dst_offset, src, src_offset, n):
...
+            types.unicode_type)(context, builder, value=src)
+        view_str = cgutils.create_struct_proxy(
+            types.unicode_type)(context, builder)
+        view_str.meminfo = in_str.meminfo
+        view_str.kind = in_str.kind
+        view_str.length = length
+        # hash value -1 to indicate "need to compute hash"
+        view_str.hash = context.get_constant(_Py_hash_t, -1)
+        # get a pointer to start of slice data
+        byte_width = context.compile_internal(
```
Comment (stuartarchibald) [inline numba/unicode.py]: Does the `resolve_value_type`, `get_call_type` etc pattern not work here?

Comment (ehsantn) [reply numba/unicode.py]: Good point. This seemed fragile to me too. Fixed.

**Comment by sklam** [comment]
It looks good.

**Comment by esc** [comment]
#3967 will fix the tests most likely.

**Comment by ehsantn** [comment]
Thanks @esc , just merged master to incorporate the fix.

**Comment by stuartarchibald** [comment]
For the future, CI builds are merge builds incorporating master, so just restarted the builds should be sufficient.

**Comment by ehsantn** [comment]
Ok, thanks.

**Comment by seibert** [comment]
This is ready to merge.

**Pull request merged**